### PR TITLE
fix bytes and bytearray fromhex to support subclass

### PR DESCRIPTION
**Pull request by qingshi163**

### CODE REVIEW

File: vm/src/builtins/bytearray.rs [hunk 1]
```diff
@@ -373,9 +375,15 @@ impl PyByteArray {
...
-    fn fromhex(string: PyStrRef, vm: &VirtualMachine) -> PyResult<PyByteArray> {
-        Ok(PyBytesInner::fromhex(string.as_str(), vm)?.into())
+    #[pyclassmethod]
+    fn fromhex(cls: PyTypeRef, string: PyStrRef, vm: &VirtualMachine) -> PyResult {
+        let bytes = PyBytesInner::fromhex(string.as_str(), vm)?;
+        let bytes = vm.ctx.new_bytes(bytes);
+        let args = FuncArgs {
+            args: vec![bytes],
+            ..Default::default()
+        };
```
Comment (qingshi163) [inline vm/src/builtins/bytearray.rs:385]: better way to get FuncArgs from Vec<u8>?

Comment (youknowone) [reply vm/src/builtins/bytearray.rs:385]: I think `FuncArgs::from(vec![..])` works

File: vm/src/builtins/bytearray.rs [hunk 2]
```diff
@@ -373,9 +375,12 @@ impl PyByteArray {
...
 
-    #[pymethod]
-    fn fromhex(string: PyStrRef, vm: &VirtualMachine) -> PyResult<PyByteArray> {
-        Ok(PyBytesInner::fromhex(string.as_str(), vm)?.into())
+    #[pyclassmethod]
+    fn fromhex(cls: PyTypeRef, string: PyStrRef, vm: &VirtualMachine) -> PyResult {
+        let bytes = PyBytesInner::fromhex(string.as_str(), vm)?;
+        let bytes = vm.ctx.new_bytes(bytes);
+        let args = FuncArgs::from(vec![bytes]);
+        Callable::call(&cls, args, vm)
```
Comment (youknowone) [inline vm/src/builtins/bytearray.rs:383]:
```suggestion
        Callable::call(&cls, vec![bytes].into(), vm)
```
Oh, I didn't look in the usage very well. maybe this expression also works

File: vm/src/builtins/bytearray.rs [hunk 3]
```diff
@@ -373,9 +375,12 @@ impl PyByteArray {
         self.inner().hex(sep, bytes_per_sep, vm)
     }
 
-    #[pymethod]
-    fn fromhex(string: PyStrRef, vm: &VirtualMachine) -> PyResult<PyByteArray> {
-        Ok(PyBytesInner::fromhex(string.as_str(), vm)?.into())
+    #[pyclassmethod]
+    fn fromhex(cls: PyTypeRef, string: PyStrRef, vm: &VirtualMachine) -> PyResult {
+        let bytes = PyBytesInner::fromhex(string.as_str(), vm)?;
+        let bytes = vm.ctx.new_bytes(bytes);
```
Comment (youknowone) [inline vm/src/builtins/bytearray.rs:381]: to check if this is intended: this is `bytearray` but calling `new_bytes`, right?

**Review by youknowone — APPROVED** [review]

File: vm/src/builtins/bytearray.rs (hunk 3 above)
Comment (qingshi163) [reply vm/src/builtins/bytearray.rs:381]: yes, see the updated test code. the bytes will pass to __new__ or __init__ in python than back to tp_new.

**Pull request merged**